### PR TITLE
[Merged by Bors] - feat: port Algebra.GCDMonoid.Multiset

### PR DESCRIPTION
**Pull request by maxwell-thum**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/GCDMonoid/Multiset.lean [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+this lower priority to avoid linter complaints about simp-normal form -/
+/- Porting note: Mathport seems to be replacing `multiset.induction_on s $` with
+`(Multiset.induction_on s)`, when it should be `Multiset.induction_on s <|`. -/
+@[simp 1100]
+theorem normalize_lcm (s : Multiset α) : normalize s.lcm = s.lcm :=
+  Multiset.induction_on s (by simp) <| fun a s _ ↦ by simp
+#align multiset.normalize_lcm Multiset.normalize_lcm
+
+/- Porting note: When trying to use `simp only [lcm_eq_zero_iff]`, Lean seems to think
+`lcm_eq_zero_iff` refers to the very theorem we are trying to prove. -/
```
Comment (jcommelin) [inline Mathlib/Algebra/GCDMonoid/Multiset.lean:89]:
```suggestion
```
This is intentional. So it doesn't need a porting note. Either fix it with `nonrec`, as you did, or by prefixing `lcm_eq_zero_iff` in the proof with a namespace.

File: Mathlib/Algebra/GCDMonoid/Multiset.lean [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+@[simp]
+nonrec theorem lcm_eq_zero_iff [Nontrivial α] (s : Multiset α) : s.lcm = 0 ↔ (0 : α) ∈ s :=
+  by
```
Comment (jcommelin) [inline Mathlib/Algebra/GCDMonoid/Multiset.lean:92]:
```suggestion
nonrec theorem lcm_eq_zero_iff [Nontrivial α] (s : Multiset α) : s.lcm = 0 ↔ (0 : α) ∈ s := by
```

File: Mathlib/Algebra/GCDMonoid/Multiset.lean [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+  · simp only [lcm_zero, one_ne_zero, not_mem_zero]
+  · simp only [mem_cons, lcm_cons, lcm_eq_zero_iff, ihs, @eq_comm _ a]
+#align multiset.lcm_eq_zero_iff Multiset.lcm_eq_zero_iff
+
+variable [DecidableEq α]
+
+@[simp]
+theorem lcm_dedup (s : Multiset α) : (dedup s).lcm = s.lcm :=
+  Multiset.induction_on s (by simp) <| fun a s IH ↦
+    by
```
Comment (jcommelin) [inline Mathlib/Algebra/GCDMonoid/Multiset.lean:103]:
```suggestion
  Multiset.induction_on s (by simp) <| fun a s IH ↦ by
```

File: Mathlib/Algebra/GCDMonoid/Multiset.lean [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+    by_cases a ∈ s <;> simp [IH, h]
+    unfold lcm
+    rw [← cons_erase h, fold_cons_left, ← lcm_assoc, lcm_same]
+    apply lcm_eq_of_associated_left (associated_normalize _)
+#align multiset.lcm_dedup Multiset.lcm_dedup
+
+@[simp]
+theorem lcm_ndunion (s₁ s₂ : Multiset α) : (ndunion s₁ s₂).lcm = GCDMonoid.lcm s₁.lcm s₂.lcm :=
+  by
```
Comment (jcommelin) [inline Mathlib/Algebra/GCDMonoid/Multiset.lean:112]:
```suggestion
theorem lcm_ndunion (s₁ s₂ : Multiset α) : (ndunion s₁ s₂).lcm = GCDMonoid.lcm s₁.lcm s₂.lcm := by
```

File: Mathlib/Algebra/GCDMonoid/Multiset.lean [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+  rw [← lcm_dedup, dedup_ext.2, lcm_dedup, lcm_add]
+  simp
+#align multiset.lcm_ndunion Multiset.lcm_ndunion
+
+@[simp]
+theorem lcm_union (s₁ s₂ : Multiset α) : (s₁ ∪ s₂).lcm = GCDMonoid.lcm s₁.lcm s₂.lcm :=
+  by
```
Comment (jcommelin) [inline Mathlib/Algebra/GCDMonoid/Multiset.lean:119]:
```suggestion
theorem lcm_union (s₁ s₂ : Multiset α) : (s₁ ∪ s₂).lcm = GCDMonoid.lcm s₁.lcm s₂.lcm := by
```

File: Mathlib/Algebra/GCDMonoid/Multiset.lean [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+  rw [← lcm_dedup, dedup_ext.2, lcm_dedup, lcm_add]
+  simp
+#align multiset.lcm_union Multiset.lcm_union
+
+@[simp]
+theorem lcm_ndinsert (a : α) (s : Multiset α) : (ndinsert a s).lcm = GCDMonoid.lcm a s.lcm :=
+  by
```
Comment (jcommelin) [inline Mathlib/Algebra/GCDMonoid/Multiset.lean:126]:
```suggestion
theorem lcm_ndinsert (a : α) (s : Multiset α) : (ndinsert a s).lcm = GCDMonoid.lcm a s.lcm := by
```

File: Mathlib/Algebra/GCDMonoid/Multiset.lean [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+
+/- Porting note: Following `Algebra.GCDMonoid.Basic`'s version of `normalize_gcd`, I'm giving
+this lower priority to avoid linter complaints about simp-normal form -/
+@[simp 1100]
+theorem normalize_gcd (s : Multiset α) : normalize s.gcd = s.gcd :=
+  Multiset.induction_on s (by simp) <| fun a s _ ↦ by simp
+#align multiset.normalize_gcd Multiset.normalize_gcd
+
+theorem gcd_eq_zero_iff (s : Multiset α) : s.gcd = 0 ↔ ∀ x : α, x ∈ s → x = 0 :=
+  by
```
Comment (jcommelin) [inline Mathlib/Algebra/GCDMonoid/Multiset.lean:184]:
```suggestion
theorem gcd_eq_zero_iff (s : Multiset α) : s.gcd = 0 ↔ ∀ x : α, x ∈ s → x = 0 := by
```

File: Mathlib/Algebra/GCDMonoid/Multiset.lean [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+    rw [← h]
+    apply gcd_dvd hx
+  · refine' s.induction_on _ _
+    · simp
+    intro a s sgcd h
+    simp [h a (mem_cons_self a s), sgcd fun x hx ↦ h x (mem_cons_of_mem hx)]
+#align multiset.gcd_eq_zero_iff Multiset.gcd_eq_zero_iff
+
+theorem gcd_map_mul (a : α) (s : Multiset α) : (s.map ((· * ·) a)).gcd = normalize a * s.gcd :=
+  by
```
Comment (jcommelin) [inline Mathlib/Algebra/GCDMonoid/Multiset.lean:197]:
```suggestion
theorem gcd_map_mul (a : α) (s : Multiset α) : (s.map ((· * ·) a)).gcd = normalize a * s.gcd := by
```

File: Mathlib/Algebra/GCDMonoid/Multiset.lean [hunk 9]
```diff
@@ -0,0 +1,279 @@
...
+#align multiset.gcd_map_mul Multiset.gcd_map_mul
+
+section
+
+variable [DecidableEq α]
+
+@[simp]
+theorem gcd_dedup (s : Multiset α) : (dedup s).gcd = s.gcd :=
+  Multiset.induction_on s (by simp) <| fun a s IH ↦
+    by
```
Comment (jcommelin) [inline Mathlib/Algebra/GCDMonoid/Multiset.lean:212]:
```suggestion
  Multiset.induction_on s (by simp) <| fun a s IH ↦ by
```

File: Mathlib/Algebra/GCDMonoid/Multiset.lean [hunk 10]
```diff
@@ -0,0 +1,279 @@
...
+    by_cases a ∈ s <;> simp [IH, h]
+    unfold gcd
+    rw [← cons_erase h, fold_cons_left, ← gcd_assoc, gcd_same]
+    apply (associated_normalize _).gcd_eq_left
+#align multiset.gcd_dedup Multiset.gcd_dedup
+
+@[simp]
+theorem gcd_ndunion (s₁ s₂ : Multiset α) : (ndunion s₁ s₂).gcd = GCDMonoid.gcd s₁.gcd s₂.gcd :=
+  by
```
Comment (jcommelin) [inline Mathlib/Algebra/GCDMonoid/Multiset.lean:221]:
```suggestion
theorem gcd_ndunion (s₁ s₂ : Multiset α) : (ndunion s₁ s₂).gcd = GCDMonoid.gcd s₁.gcd s₂.gcd := by
```

File: Mathlib/Algebra/GCDMonoid/Multiset.lean [hunk 11]
```diff
@@ -0,0 +1,279 @@
...
+  rw [← gcd_dedup, dedup_ext.2, gcd_dedup, gcd_add]
+  simp
+#align multiset.gcd_ndunion Multiset.gcd_ndunion
+
+@[simp]
+theorem gcd_union (s₁ s₂ : Multiset α) : (s₁ ∪ s₂).gcd = GCDMonoid.gcd s₁.gcd s₂.gcd :=
+  by
```
Comment (jcommelin) [inline Mathlib/Algebra/GCDMonoid/Multiset.lean:228]:
```suggestion
theorem gcd_union (s₁ s₂ : Multiset α) : (s₁ ∪ s₂).gcd = GCDMonoid.gcd s₁.gcd s₂.gcd := by
```

File: Mathlib/Algebra/GCDMonoid/Multiset.lean [hunk 12]
```diff
@@ -0,0 +1,279 @@
...
+  rw [← gcd_dedup, dedup_ext.2, gcd_dedup, gcd_add]
+  simp
+#align multiset.gcd_union Multiset.gcd_union
+
+@[simp]
+theorem gcd_ndinsert (a : α) (s : Multiset α) : (ndinsert a s).gcd = GCDMonoid.gcd a s.gcd :=
+  by
```
Comment (jcommelin) [inline Mathlib/Algebra/GCDMonoid/Multiset.lean:235]:
```suggestion
theorem gcd_ndinsert (a : α) (s : Multiset α) : (ndinsert a s).gcd = GCDMonoid.gcd a s.gcd := by
```

File: Mathlib/Algebra/GCDMonoid/Multiset.lean [hunk 13]
```diff
@@ -0,0 +1,279 @@
...
+  simp
+#align multiset.gcd_ndinsert Multiset.gcd_ndinsert
+
+end
+
+theorem extract_gcd' (s t : Multiset α) (hs : ∃ x, x ∈ s ∧ x ≠ (0 : α))
+    (ht : s = t.map ((· * ·) s.gcd)) : t.gcd = 1 :=
+  ((@mul_right_eq_self₀ _ _ s.gcd _).1 <| by
+        conv_lhs => rw [← normalize_gcd, ← gcd_map_mul, ← ht]).resolve_right <|
+    by
```
Comment (jcommelin) [inline Mathlib/Algebra/GCDMonoid/Multiset.lean:246]:
```suggestion
        conv_lhs => rw [← normalize_gcd, ← gcd_map_mul, ← ht]).resolve_right <| by
```

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

File: Mathlib/Data/Multiset/Basic.lean [hunk 14]
```diff
@@ -1189,6 +1189,10 @@ theorem map_replicate (f : α → β) (k : ℕ) (a : α) : (replicate k a).map f
   simp only [← coe_replicate, coe_map, List.map_replicate]
 #align multiset.map_replicate Multiset.map_replicate
 
+-- Porting note: Ad hoc lemma for `Algebra.GCDMonoid.Multiset`
+theorem map_replicate (f : α → β) (k : ℕ) (a : α) : (replicate k a).map f = replicate k (f a) :=
+  map_repeat f a k
+
```
Comment (jcommelin) [inline Mathlib/Data/Multiset/Basic.lean:1195]:
```suggestion
```

**Comment by jcommelin** [comment]
bors d+

**Comment by maxwell-thum** [comment]
> ✌️ maxwell-thum can now approve this pull request. To approve and merge a pull request, simply reply with `bors r+`. More detailed instructions are available [here](https://bors.tech/documentation/getting-started/#reviewing-pull-requests).

bors r+

**Pull request closed without merging**